src/nupic/algorithms/AnomalyLikelihood.cpp

```diff
@@ -0,0 +1,592 @@
+# ~~~~~~
+#
+# The parameters dict returned by the above functions has the following
+# structure. Note: the client does not need to know the details of this.
+#
+# ::
+#
+#  {
+#    "distribution":               # describes the distribution
+#      {
+#        "name": STRING,           # name of the distribution, such as 'normal'
+#        "mean": SCALAR,           # mean of the distribution
+#        "variance": SCALAR,       # variance of the distribution
+#
+#        # There may also be some keys that are specific to the distribution
+#      },
+#
+#    "historicalLikelihoods": []   # Contains the last windowSize likelihood
+#                                  # values returned
+#
+#    "movingAverage":              # stuff needed to compute a rolling average
+#                                  # of the anomaly scores
+#      {
+#        "windowSize": SCALAR,     # the size of the averaging window
+#        "historicalValues": [],   # list with the last windowSize anomaly
+#                                  # scores
+#        "total": SCALAR,          # the total of the values in historicalValues
+#      },
+#
+#  }
+
+**/
+  /** returns parameters of a Null distribution **/
+  DistributionParams nullDistribution() const {
+    return DistributionParams("normal", 0.5, 1e6, 1e3);
+  }
+
+ Real32 tailProbability(Real32 x) const {
+ /**
+  Given the normal distribution specified by the mean and standard deviation
+  in distributionParams, return the probability of getting samples further
+  from the mean. For values above the mean, this is the probability of getting
+  samples > x and for values below the mean, the probability of getting
+  samples < x. This is the Q-function: the tail probability of the normal distribution.
+
+  :param distributionParams: dict with 'mean' and 'stdev' of the distribution
+  **/
+     assert(distribution.name != "unknown" && distribution.stdev > 0);
+     
+  if (x < distribution.mean) {
+    // Gaussian is symmetrical around mean, so flip to get the tail probability
+    Real32 xp = 2 * distribution.mean - x;
+    assert(xp != x);
+    return tailProbability(xp);
+  }
+
+  // Calculate the Q function with the complementary error function, explained
+  // here: http://www.gaussianwaves.com/2012/07/q-function-and-error-functions
+  Real32 z = (x - distribution.mean) / distribution.stdev;
+  return 0.5 * erfc(z/1.4142);
+  }
+
+
+inline Real  computeLogLikelihood(Real likelihood) const {
+  /**
+    Compute a log scale representation of the likelihood value. Since the
+    likelihood computations return low probabilities that often go into four 9's
+    or five 9's, a log value is more useful for visualization, thresholding,
+    etc.
+   **/
+    // The log formula is:
+    //     Math.log(1.0000000001 - likelihood) / Math.log(1.0 - 0.9999999999)
+    return log(1.0000000001f - likelihood) / -23.02585084720009f;
+}
+
+
+DistributionParams estimateNormal(vector<Real> sampleData, bool performLowerBoundCheck=true) {
```

Comment:
TODO: should I create a class DistributionParams (currently a struct) and check/compute all this info in the class, instead of this method?